### DISCUSSION TITLE
Problem with req. query

### DISCUSSION THREAD

**Discussion by dragoslavIvkovic** [vercel/next.js#38584] [Help]
I'm trying to pull data from a document containing a logged-in user from the MongoDB database. However, req. query doesn't seem to pass.

      `import clientPromise from "../../lib/mongodb";
      
      export default async function handler(req, res) {
        const client = await clientPromise;
        const db = client.db("USERS");
      
        switch (req.method) {
        case "POST": {
          const bodyObject = JSON.parse(req.body);
          const query = {
            user: bodyObject.user,
            avatar: bodyObject.avatar,
            level: bodyObject.level,
          };
          const update = { $set: { score: bodyObject.score } };
          const options = { upsert: true };
      
          const newScore = await db
            .collection("USERS")
            .findOneAndUpdate(query, update, options);
      
          res.json(newScore);
          break;
        }
        case "GET": {
          const { userName } = req.query;
          const users = await db
            .collection("USERS")
            .find({ user: { $in: [userName] } })
            .toArray();
          res.json(users);
          break;
        }
      
        default: {
          // do nothing
        }
        }
      }
      
      
      import styles from '../styles/Elements.module.css';
      import clientPromise from '../lib/mongodb';
       
      
      export default function ScoreHeader({ users }) {
        console.log(users)
        
        return (
          <div className={styles.containerIndex}>
           fadsfds
          </div>
        );
      }
      
      export async function getServerSideProps({}) {
        const client = await clientPromise;
       const db = client.db('USERS');
       
        const res = await fetch(`http://localhost:3000/api/usersAPI`, {
          method: 'GET',
          headers: {
            'Content-Type': 'application/json',
          },
        });
        const users = await res.json();
      
        return {
          props: { users },
        };
      }
      
      `

**Comment by icyJoseph** [comment]
Is your expectation that the call from `getServerSideProps` to `/api/usersAPI`, should contain the request query from the browser/visitor?

That's not correct, at all. And in some way, it is part of a bigger problem here. From the [docs](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes):

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

So the root issue here is that when a request arrives at `getServerSideProps`, instead of using the context of that request, as inputs, to get the data from MongoDB directly, you start a new jump toward `/api/usersAPI`, which is not necessary and creates these type of pitfalls.

The solution would be extract the query:

```ts
export async function getServerSideProps(context) {
    const client = await clientPromise;
   const db = client.db('USERS');
   const {query} = context;

   const pathname = `http://localhost:3000/api/usersAPI?userName=${query.userName}`
   
    const res = await fetch(pathname, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    const users = await res.json();
  
    return {
      props: { users },
    };
  }
```

However, that's an anti-pattern, and the need to pick on query and forward that it's a symptom. Another problem is that this would only work for `localhost:3000`.

Instead, you could define:

```js
const findByUserName = async (db, { userName }) => {
  const users = await db
    .collection("USERS")
    .find({ user: { $in: [userName] } })
    .toArray();
  return users;
};
```

And use it in `getServerSideProps`, or `api/userName`:

```js
 export async function getServerSideProps(context) {
    const client = await clientPromise;
    const db = client.db('USERS');
    const userName = context.query.userName;
    const users = await findByUserName(db, {userName}); // probably should check that it is present in the query
  
    return {
      props: { users },
    };
  }

// and in the your API route as you have it:

 case "GET": {
      const { userName } = req.query;
      const users = await findByUserName(db, {userName});
      res.json(users); // might be better to just return...
      break;
```

And now it is independent of the localhost url and it doesn't create extra round trips.


Feel free to ask more if you have any doubts.
